src/main/java/net/logstash/logback/pattern/AbstractJsonPatternParser.java

```diff
@@ -259,323 +170,380 @@ protected Object transform(final String value) throws IOException {
         }
-
-        @Override
-        public boolean shouldWrite(JsonGenerator generator, Event event) {
-            return !omitEmptyFields
-                    || (value != null
-                            && !(value instanceof JsonNode && ((JsonNode) value).getNodeType() == JsonNodeType.NULL));
+        
+        PatternLayoutAdapter<Event> layout = buildLayout(data);
+        
+        /*
+         * If layout is constant, get the constant value immediately to avoid rendering it into
+         * a StringBuilder at runtime every time an event is serialized
+         */
+        if (layout.isConstant()) {
+            final String constantValue = layout.getConstantValue();
+            return g -> constantValue;
+        } else {
+            return new LayoutValueGetter<>(layout);
         }
     }
-
-    protected class ListWriter implements NodeWriter<Event> {
-        private final List<NodeWriter<Event>> items;
-
-        public ListWriter(final List<NodeWriter<Event>> items) {
-            this.items = items;
+    
+    
+    protected PatternLayoutAdapter<Event> buildLayout(String format) {
+        PatternLayoutBase<Event> layout = createLayout();
+        
+        if (layout.isStarted()) {
+            throw new IllegalStateException("PatternLayout should not be started");
         }
+        
+        layout.setContext(contextAware.getContext());
+        layout.setPattern(format);
+        layout.setPostCompileProcessor(null); // Remove EnsureLineSeparation which is there by default
+       
+        return new PatternLayoutAdapter<>(layout);
+    }
 
-        public void write(JsonGenerator generator, Event event) throws IOException {
-            generator.writeStartArray();
-            for (NodeWriter<Event> item : items) {
-                if (item.shouldWrite(generator, event)) {
-                    item.write(generator, event);
-                }
-            }
-            generator.writeEndArray();
+    
+    /**
+     * Create a PatternLayout instance of the appropriate type. The returned instance
+     * will further configured with the context and appropriate pattern then started.
+     * 
+     * @return an unstarted {@link PatternLayoutBase} instance
+     */
+    protected abstract PatternLayoutBase<Event> createLayout();
+    
+    
+    protected static class LayoutValueGetter<Event> implements ValueGetter<String, Event> {
+        /**
+         * The PatternLayout from which the value is generated
+         */
+        private final PatternLayoutAdapter<Event> layout;
+
+        /**
+         * ThreadLocal reusable StringBuilder instances
+         */
+        private static final ThreadLocal<StringBuilder> stringBuilders = ThreadLocal.withInitial(StringBuilder::new);
+        
+        /**
+         * StringBuilder whose length after use exceeds the maxRecylableSize will be
+         * discarded instead of recycled.
+         */
+        private static final int maxRecyclableSize = 2048;
```

Comment:
MAX_RECYCLABLE_SIZE

Reply:
Ok with the case.
The idea here is to have a maximum length and avoid keeping very large string builders around (may be produced when rendering stack traces for instance). Considering that the string builder is used to render a single field of the pattern at once, `2048` is already a (very) large buffer... `1024` could be enough... what you think?

Reply:
1024 seems good